SDWebImage/UIView+WebCache.m

```diff
@@ -52,31 +52,57 @@ - (void)sd_internalSetImageWithURL:(nullable NSURL *)url
         
         __weak __typeof(self)wself = self;
         id <SDWebImageOperation> operation = [SDWebImageManager.sharedManager loadImageWithURL:url options:options progress:progressBlock completed:^(UIImage *image, NSData *data, NSError *error, SDImageCacheType cacheType, BOOL finished, NSURL *imageURL) {
+            // in fact, this completed block is **always** called from main queue. Maybe we should do some cleanup
             __strong __typeof (wself) sself = wself;
             [sself sd_removeActivityIndicator];
             if (!sself) {
                 return;
             }
-            dispatch_main_async_safe(^{
-                if (!sself) {
-                    return;
-                }
-                if (image && (options & SDWebImageAvoidAutoSetImage) && completedBlock) {
-                    completedBlock(image, error, cacheType, url);
-                    return;
-                } else if (image) {
-                    [sself sd_setImage:image imageData:data basedOnClassOrViaCustomSetImageBlock:setImageBlock];
-                    [sself sd_setNeedsLayout];
-                } else {
-                    if ((options & SDWebImageDelayPlaceholder)) {
-                        [sself sd_setImage:placeholder imageData:nil basedOnClassOrViaCustomSetImageBlock:setImageBlock];
+            if ((image && (options & SDWebImageAvoidAutoSetImage)) || (!image && !(options & SDWebImageDelayPlaceholder))) {
+                dispatch_main_async_safe(^{
+                    if (completedBlock) {
+                        completedBlock(image, error, cacheType, url);
+                    }
+                });
+                // no need to call setImageBlock and return
+                return;
+            }
+            
+            UIImage *targetImage;
+            NSData *targetData;
+            if (image) {
+                targetImage = image;
+                targetData = data;
+            } else if (options & SDWebImageDelayPlaceholder) {
+                targetImage = placeholder;
+                targetData = nil;
+            }
+            if ([sself isKindOfClass:NSClassFromString(@"FLAnimatedImageView")]) {
```

Comment:
what does this check do when we don't have the GIF component (including FLAnimatedImageView). Maybe we can check `if ([sself isMemberOfClass:[UIImageView class]])` ?

Reply:
This is ensure only `FLAnimatedImageView` will enter this logic (global queue setImageBlock). The else condition is for all-other situation.

Reply:
- I was referring to installing the lib without GIF (`pod 'SDWebImage/core'`). Then the `FLAnimatedImageView` class does not exist.
- the check I suggest was for `UIImageView`, so if it's exactly `UIImageView`, then execute on main queue, otherwise on the background queue. But It's not such a good idea

Reply:
I can't say that I like `isKindOfClass` check but it seems to the only way. Also If we are using `FLAnimatedImage` in so many places, we don't need to pretend that we are not using it, even when "installing the lib without GIF".

Reply:
@bpoplauschi @skyline75489 Why this is a little hack because we do not have something like `#if SD_WEP` such as `#if SD_GIF`(Maybe we can add one). Use `#if __has_include` and check that class is OK but looks too complicated. So currently I use that `NSClassFromString`. Any good idea.?

Reply:
@dreampiggy I feel that currently the GIF-related code is somehow tight-coupled with the core code. The reason for this perhaps lays within the nature of GIF format itself. It's just combination of simple images. The decoupling of GIF code may not be intuitive or the right thing to do.

Reply:
@skyline75489 Agreed. Maybe we need to refacotor some code. Currently our `sd_internalSetImageWithURL` is not too customizable. I advice too add a extra context arg. And another suggestion is to move this code into a actually `private` header files. (Through cocoapods's `private_files`) to avoid our user touch it or even relay on it (In fact, my company's a little code relay on this method 😂).

Reply:
Through context arg. We can pass extra information and let the caller to do specify logic. For example, if we can get `call_setImageBlock_in_this_queue` from the context. Then we do this logic. The context arg should be set from `FLAnimatedImageView+WebCache`.

A context can be a naive `NSDictionary` and we grab things through some private-defined-keys. Or we can make it a `id` or something else.(This is not really important)

Reply:
@skyline75489 please detail on * I feel that currently the GIF-related code is somehow tight-coupled with the core code.*

@dreampiggy indeed we can pass something for more customisation. I would use a dictionary of string,string like `options` dict for `decompressedImageWithImage`
That way we could pass options like how to run the `setImageBlock` which would move the decision to the `FLAnimatedImageView` category - much cleaner.